### PR TITLE
Feature/serialization

### PR DESCRIPTION
**Pull request by xkevin190**
- serialization was added to the messages
- message compression was added

### CODE REVIEW

File: android/app/src/main/rust/build.rs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+extern crate prost_build;
+
+fn main() {
+    prost_build::compile_protos(&["src/message.proto"],
+                                &["src/"]).unwrap();
```
Comment (jeandudey) [inline android/app/src/main/rust/build.rs:5]:
```suggestion
    prost_build::compile_protos(&["src/message.proto"]).unwrap();
```

File: android/app/src/main/rust/src/runtime.rs [hunk 2]
```diff
@@ -43,453 +44,593 @@ use crate::util::{unwrap_exc_or, unwrap_exc_or_default, unwrap_jni};
...
+  "Lio/locha/p2p/runtime/RuntimeEvents;";
+
+pub mod items {
+  include!(concat!(env!("OUT_DIR"), "/message.items.rs"));
+}
+
+#[derive(Serialize, Deserialize)]
+struct Msg {
+  text: String,
+  typeFile: Option<String>,
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:56]:
```suggestion
  #[serde(rename = "typeFile")]
  type_file: Option<String>,
```

File: android/app/src/main/rust/src/runtime.rs [hunk 3]
```diff
@@ -43,453 +44,593 @@ use crate::util::{unwrap_exc_or, unwrap_exc_or_default, unwrap_jni};
...
+  file: Option<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct DataStatus {
+  status: String,
+  msgID: String,
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:63]:
```suggestion
  #[serde(rename = "msgID")]
  msg_id: String,
```

File: android/app/src/main/rust/src/runtime.rs [hunk 4]
```diff
@@ -43,453 +44,593 @@ use crate::util::{unwrap_exc_or, unwrap_exc_or_default, unwrap_jni};
...
+}
+
+#[derive(Serialize, Deserialize)]
+struct Person {
+  toUID: String,
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:68]:
```suggestion
  #[serde(rename = "toUID")]
  to_uid: String,
```

File: android/app/src/main/rust/src/runtime.rs [hunk 5]
```diff
@@ -43,453 +44,593 @@ use crate::util::{unwrap_exc_or, unwrap_exc_or_default, unwrap_jni};
...
+  msgID: String,
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:69]:
```suggestion
  #[serde(rename = "msgID")]
  msgID: String,
```

File: android/app/src/main/rust/src/runtime.rs [hunk 6]
```diff
@@ -43,453 +44,593 @@ use crate::util::{unwrap_exc_or, unwrap_exc_or_default, unwrap_jni};
...
+}
+
+#[derive(Serialize, Deserialize)]
+struct DataStatus {
+  status: String,
+  msgID: String,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Person {
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:67]: Person -> Message :wink:

File: android/app/src/main/rust/src/runtime.rs [hunk 7]
```diff
@@ -43,453 +44,593 @@ use crate::util::{unwrap_exc_or, unwrap_exc_or_default, unwrap_jni};
...
+  toUID: String,
+  msgID: String,
+  timestamp: u64,
+  shippingTime: Option<u64>,
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:71]:
```suggestion
  #[serde(rename = "shippingTime")]
  shipping_time: Option<u64>,
```

File: android/app/src/main/rust/src/runtime.rs [hunk 8]
```diff
@@ -43,453 +44,593 @@ use crate::util::{unwrap_exc_or, unwrap_exc_or_default, unwrap_jni};
...
+  r#type: u32,
+  msg: Msg,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Status {
+  toUID: String,
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:78]:
```suggestion
  #[serde(rename = "toUID")]
  to_uid: String,
```

File: android/app/src/main/rust/src/runtime.rs [hunk 9]
```diff
@@ -43,453 +44,593 @@ use crate::util::{unwrap_exc_or, unwrap_exc_or_default, unwrap_jni};
...
+      message.text = person.msg.text;
+      message.file = person.msg.file.unwrap_or(String::new());
+      message.type_file = person.msg.typeFile.unwrap_or(String::new());
+    }
+    ContentMessage::Status(status) => {
+      message.to_uid = status.toUID;
+      message.timestamp = status.timestamp;
+      message.type_message = status.r#type;
+      message.msg_id = status.data.msgID;
+      message.status = status.data.status
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:255]:
```suggestion
      message.status = status.data.status;
```

File: android/app/src/main/rust/src/runtime.rs [hunk 10]
```diff
@@ -43,453 +44,593 @@ use crate::util::{unwrap_exc_or, unwrap_exc_or_default, unwrap_jni};
...
-        task::block_on(runtime.send_message(contents));
-        Ok(())
-    });
-    unwrap_exc_or_default(&env, res)
+  trace!("nativeSendMessage");
+
+  let res = panic::catch_unwind(|| {
+    let runtime = env.get_rust_field::<_, _, Runtime>(class, "handle")?;
+    let contents: String = env.get_string(contents)?.into();
+    let serilize: Vec<u8> = serialize_message(&contents);
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:331]:
```suggestion
    let serialized: Vec<u8> = serialize_message(&contents);
```

**Review by jeandudey — CHANGES_REQUESTED** [review]
Please also run `cargo fmt --all` rust source code directory :wink:

File: android/app/src/main/rust/src/runtime.rs (hunk 6 above)
Comment (xkevin190) [reply android/app/src/main/rust/src/runtime.rs]: Message is already being used and there is an error, rename it MessageData

File: android/app/src/main/rust/src/runtime.rs [hunk 11]
```diff
@@ -200,7 +337,8 @@ pub extern "system" fn Java_io_locha_p2p_runtime_Runtime_nativeSendMessage(
     let res = panic::catch_unwind(|| {
         let runtime = env.get_rust_field::<_, _, Runtime>(class, "handle")?;
         let contents: String = env.get_string(contents)?.into();
-        task::block_on(runtime.send_message(contents));
+        let serilized: Vec<u8> = serialize_message(&contents);
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:340]:
```suggestion
        let serialized: Vec<u8> = serialize_message(&contents);
```

File: android/app/src/main/rust/src/runtime.rs [hunk 12]
```diff
@@ -219,19 +357,18 @@ impl RuntimeEventsProxy {
 }
 
 impl RuntimeEvents for RuntimeEventsProxy {
-    fn on_new_message(&mut self, peer_id: &PeerId, message: String) {
+    fn on_new_message(&mut self, peer_id: &PeerId, message: Vec<u8>) {
         unwrap_jni(self.exec.with_attached(|env| {
+            let str_message: String = deserialize_message(&message);
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:362]:
```suggestion
            let message_json = deserialize_message(&message);
```

File: android/app/src/main/rust/src/runtime.rs [hunk 13]
```diff
@@ -189,6 +237,95 @@ pub extern "system" fn Java_io_locha_p2p_runtime_Runtime_nativeDial(
     unwrap_exc_or_default(&env, res)
 }
 
+pub fn serialize_message(contents: &String) -> Vec<u8> {
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:240]:
```suggestion
pub fn serialize_message(contents: String) -> Vec<u8> {
```

File: android/app/src/main/rust/src/runtime.rs (hunk 11 above)
Comment (xkevin190) [reply android/app/src/main/rust/src/runtime.rs:340]: it has been already added

File: android/app/src/main/rust/src/runtime.rs [hunk 14]
```diff
@@ -189,6 +237,95 @@ pub extern "system" fn Java_io_locha_p2p_runtime_Runtime_nativeDial(
...
+    trace!("json string {}", contents);
+
+    let mut buf = Vec::new();
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:243]: Let's move it to where it's used (on the encoder)

File: android/app/src/main/rust/src/runtime.rs [hunk 15]
```diff
@@ -189,6 +237,95 @@ pub extern "system" fn Java_io_locha_p2p_runtime_Runtime_nativeDial(
...
+
+pub fn deserialize_message(buf: &[u8]) -> String {
+    let mut decode = Decoder::new();
+    let decompress_bytes =
+        decode.decompress_vec(buf).expect("decompress failed");
+
+    let content: items::Content =
+        items::Content::decode(&mut Cursor::new(&decompress_bytes)).unwrap();
+
+    trace!("status123 {}", content.status);
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:286]:
```suggestion
```

File: android/app/src/main/rust/src/runtime.rs [hunk 16]
```diff
@@ -45,6 +46,53 @@ use crate::{JniError, JniErrorKind};
...
+    timestamp: u64,
+    #[serde(rename = "shippingTime")]
+    shipping_time: Option<u64>,
+    r#type: u32,
+    msg: Msg,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Status {
+    toUID: String,
```
Comment (jeandudey) [inline android/app/src/main/rust/src/runtime.rs:83]:
```suggestion
   #[serde(rename = "toUID")]
    to_uid: String,
```

**Review by jeandudey — APPROVED** [review]
Only one minor nit, after it's solved we can merge. :tada:

**Review by jeandudey — APPROVED** [review]

**Review by luisan00 — APPROVED** [review]

**Pull request merged**